test/WebJobs.Script.Tests/Description/Worker/WorkerFunctionInvokerTests.cs

```diff
@@ -0,0 +1,87 @@
+
+using System;
+using System.Collections.ObjectModel;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Script.Binding;
+using Microsoft.Azure.WebJobs.Script.Description;
+using Microsoft.Azure.WebJobs.Script.Workers;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.WebJobs.Script.Tests;
+using Moq;
+using Xunit;
+
+namespace Microsoft.Azure.WebJobs.Script.Tests
+{
+    public class WorkerFunctionInvokerTests
+    {
+        private readonly TestWorkerFunctionInvoker _testFunctionInvoker;
+        private readonly Mock<IScriptJobHostEnvironment> _mockJobHostEnvironment;
+        private readonly Mock<IFunctionInvocationDispatcher> _mockFunctionInvocationDispatcher;
+
+        public WorkerFunctionInvokerTests()
+        {
+            _mockJobHostEnvironment = new Mock<IScriptJobHostEnvironment>();
+            _mockFunctionInvocationDispatcher = new Mock<IFunctionInvocationDispatcher>();
+            _mockFunctionInvocationDispatcher.Setup(a => a.ErrorEventsThreshold).Returns(0);
+
+            var hostBuilder = new HostBuilder()
+                .ConfigureDefaultTestWebScriptHost(o =>
+                {
+                    o.ScriptPath = TestHelpers.FunctionsTestDirectory;
+                    o.LogPath = TestHelpers.GetHostLogFileDirectory().Parent.FullName;
+                });
+            var host = hostBuilder.Build();
+
+            var sc = host.GetScriptHost();
+
+            FunctionMetadata metaData = new FunctionMetadata();
+            _testFunctionInvoker = new TestWorkerFunctionInvoker(sc, null, metaData, NullLoggerFactory.Instance, null, new Collection<FunctionBinding>(), _mockFunctionInvocationDispatcher.Object, _mockJobHostEnvironment.Object);
+        }
+
+        [Fact]
+        public async Task InvokeTimeout_CallsShutdown()
+        {
+            try
+            {
+                _mockFunctionInvocationDispatcher.Setup(a => a.State).Returns(FunctionInvocationDispatcherState.Initializing);
+                await Task.WhenAny(_testFunctionInvoker.InvokeCore(new object[] { }, null), Task.Delay(TimeSpan.FromSeconds(500)));
+            }
+            catch (Exception)
+            {
+            }
+            _mockJobHostEnvironment.Verify(a => a.Shutdown(), Times.Once);
+        }
+
+        [Fact]
+        public async Task InvokeInitialized_DoesNotCallShutdown()
+        {
+            try
+            {
+                _mockFunctionInvocationDispatcher.Setup(a => a.State).Returns(FunctionInvocationDispatcherState.Initialized);
+                await Task.WhenAny(_testFunctionInvoker.InvokeCore(new object[] { }, null), Task.Delay(TimeSpan.FromSeconds(125)));
+            }
+            catch (Exception)
+            {
+            }
+            _mockJobHostEnvironment.Verify(a => a.Shutdown(), Times.Never);
+        }
+
+        private class TestWorkerFunctionInvoker : WorkerFunctionInvoker
```

Comment:
make this class public and move it into its own file to allow reusing TestWorkerFunctionInvoker for other tests if needed

Reply:
I could only make it internal since WorkerFunctionInvoker is internal..